tests/unit/test_ldap_backend.py

```diff
@@ -194,38 +252,120 @@ def test_search_valid_username_invalid_group(self):
                 'simple_bind_s',
                 'whoami_s',
-                'search_s',
+                'search',
+                'result',
+                'result',
                 'initialize',
                 'simple_bind_s',
                 'whoami_s',
                 'unbind',
-                'search_s',
+                'search',
+                'result',
                 'unbind'
             ]
         )
         self.assertFalse(result)
 
-
-def _do_simple_bind(uri, bind_dn, bind_pw, user_search=None, group_search=None, username=None, password=None):
-    backend = LDAPAuthenticationBackend(uri, use_tls=False, bind_dn=bind_dn, bind_pw=bind_pw, user=user_search, group=group_search)
+    def test_search_with_reference_result(self):
+        # This is for returning the referral object at calling 'result' method of LDAPObject
+        self.mock_referral = [
+            (None, ['ldap://fakeldap2.example.com/ou=cyberdyne,dc=example,dc=com']),
+        ]
+
+        user = {
+            "base_dn": "ou=users,dc=example,dc=com",
+            "search_filter": "(uid={username})",
+            "scope": "subtree",
+        }
+
+        self._is_maximum_referral_hop_exceeded = False
+        def side_effect(_):
+            self._is_maximum_referral_hop_exceeded = True
+
+        with mock.patch('st2auth_ldap_backend.ldap_backend.LOG.warning', mock.Mock(side_effect=side_effect)):
+            # This is a case that maximum referral hop will be exceeded
+            result = _do_simple_bind('', '',
+                                     user_search=user, group_search=None,
+                                     username='john_connor', password='HastaLavista',
+                                     ref_hop_limit=1)
+
+            self.assertEquals(self.ldapobj.methods_called(),[
+                    'initialize',
+                    'simple_bind_s',
+                    'whoami_s',
+                    'search',
+                    'result',
+                    'result',
+                    'result',
+                    'whoami_s',
+                    'unbind'
+                ]
+            )
+            self.assertTrue(result)
+            self.assertFalse(self._is_maximum_referral_hop_exceeded)
+
+    def test_search_with_reference_result_but_exceeded_maximum_referal_hop(self):
+        # This is for returning the referral object at calling 'result' method of LDAPObject
+        self.mock_referral = [
+            (None, ['ldap://fakeldap2.example.com/ou=cyberdyne,dc=example,dc=com']),
+        ]
+
+        user = {
+            "base_dn": "ou=users,dc=example,dc=com",
+            "search_filter": "(uid={username})",
+            "scope": "subtree",
+        }
+
+        self._is_maximum_referral_hop_exceeded = False
+        def side_effect(_):
+            self._is_maximum_referral_hop_exceeded = True
+
+        with mock.patch('st2auth_ldap_backend.ldap_backend.LOG.warning', mock.Mock(side_effect=side_effect)):
```

Comment:
This mock is too fragile. What if there's another LOG.warning that's unrelated to the referral in the module? Can you have the logger output to a stream which then you can search the stream for the message `Referral hop limit is exceeded`?